GPU rendering & anisotropic rendering

philippotto opened the pull request:
### Mailable description of changes:
 - WebKnossos supports anisotropic datasets now.

## PR Summary:
This PR includes #2402 which I closed since a lot changed, so reviewing that is not really sensible anymore. In addition to the changes introduced by #2402, this PR includes:
- Anisotropic datasets are supported now in orthogonal mode.
- There is no distinction between data textures for normal rendering and for fallback rendering. Instead one TextureBucketManager instance is responsible for managing buckets as well as fallback buckets. As a result, there are no black flashes anymore when switching between zoomsteps. When zooming out, the old fallback buckets become the new "normal" buckets. When zooming in, the old "normal" buckets become the new "fallback" buckets. That way, there are usually buckets available for the gpu to render immediately when switching zoomsteps.
- One texture bucket manager can hold multiple data textures if the hardware (e.g., iPad) does not support 8192**2 textures. Buckets are distributed across these data textures. As for the look up, there is still only one look up texture per TextureBucketManager which points to the correct index in the correct data structure using one float.

### Steps to test:
https://anisotropy.webknossos.xyz

- do normal view/trace/volume tracing actions on:
  - classic grayscale datasets
  - RGB datasets
  - multi-channel datasets
  - datasets with segmentation
  - anisotropic datasets

### Issues:
- fixes #2295 

### Todo:
- [x] Add constraint checks dependent on hardware and provided dataset/tracing. E.g., a dataset with 4 different layers cannot be rendered on iPad since it only supports 8 textures. Also show a proper warning in that case. Maybe we can also make trade-offs (e.g., by not supporting fallback rendering).
- [X] Re-add support for mappings.
- [x] Test wk under different hardware
    - [x] iPad (initial data is there but afterwards not refreshed)
    - [X] Safari crashes
    - [x] Laptop with 8 GB RAM and no dedicated GPU suffers from performance issues
- [x] Test ND Store
- Other issues:
  - [X] volume labeling in zoomStep = 1 has issues (upsampling of data seems wonky)
  - [x] Fallback rendering for anisotropic data misses some data in xz and yz planes (not very high-pri since it only affects anisotropic data and fallback rendering is not that crucial anyway).

------
- [x] Ready for review

hotzenklotz commented on app/assets/javascripts/test/model/binary/cube.spec.js:
```diff
@@ -29,7 +29,10 @@ mockRequire("libs/toast", { error: _.noop });
 test.beforeEach(t => {
-  const cube = new Cube([100, 100, 100], 3, 24);
+  const mockedLayer = {
+    resolutions: [[1, 1, 1], [2, 2, 2], [4, 4, 4], [8, 8, 8], [16, 16, 16], [32, 32, 32]],
```
isn't this the same as in `/fixtures/resolutions.js`?

hotzenklotz commented on app/assets/javascripts/test/model/binary/layers/wk_layer.spec.js:
```diff
@@ -33,6 +33,7 @@ const layerInfo = {
   name: "layername",
   category: "color",
   elementClass: "uint16",
+  resolutions: [[1, 1, 1], [2, 2, 2], [4, 4, 4], [8, 8, 8], [16, 16, 16], [32, 32, 32]],
```
same here

hotzenklotz commented on app/assets/javascripts/oxalis/model/binary/texture_bucket_manager.js:
```diff
@@ -0,0 +1,278 @@
+import { createUpdatableTexture } from "oxalis/geometries/materials/abstract_plane_material_factory";
+import { getRenderer } from "oxalis/controller/renderer";
+
+// A TextureBucketManager instance is responsible for making buckets of
```
wrong `of`

hotzenklotz commented on app/assets/javascripts/oxalis/model/binary/texture_bucket_manager.js:
```diff
@@ -0,0 +1,278 @@
+import { getRenderer } from "oxalis/controller/renderer";
+
+// A TextureBucketManager instance is responsible for making buckets of
+// avaible to the GPU.
```
typo `available`

hotzenklotz commented on app/assets/javascripts/oxalis/model/binary/ping_strategy.js:
```diff
@@ -69,12 +72,43 @@ export class PingStrategy extends AbstractPingStrategy {
+
+    let queueItemsForFallbackZoomStep = [];
+    const fallbackZoomStep = Math.min(this.cube.MAX_UNSAMPLED_ZOOM_STEP, currentZoomStep + 1);
+    // todo
```
todo?

hotzenklotz commented on app/assets/javascripts/oxalis/model/binary.js:
```diff
@@ -119,15 +117,166 @@ class Binary {
-    this.cube.on({
-      newMapping: () => this.forcePlaneRedraw(),
-    });
+    // todo
```
todo? Or remove?

hotzenklotz commented on app/assets/javascripts/oxalis/model/binary.js:
```diff
@@ -119,15 +117,166 @@ class Binary {
+  }
+
+  getTextureSize(): number {
+    // return 4096;
```
remove comment

hotzenklotz commented:
Stuff I noticed during testing:
- When zooming in I have white flash / flickering. Zooming out works, however.
![flickering](https://user-images.githubusercontent.com/1105056/38088432-14ddf3c8-335c-11e8-8d5a-6e89dea21f00.gif)
- Opening dataset `e2006_knossos` in any mode is broken. Perhaps, my local dataset is not configured correctly. See stack trace:
![image](https://user-images.githubusercontent.com/1105056/38088506-55dd1f2a-335c-11e8-92b3-105abb5bc2ca.png)
- in some cases the data does not load "all the way" and shows just the cross section:
![image](https://user-images.githubusercontent.com/1105056/38088661-db2c5786-335c-11e8-9d13-2295f797e49e.png)
- flight mode is broken
![image](https://user-images.githubusercontent.com/1105056/38088745-318e8310-335d-11e8-9002-1f488061d5b5.png)
- Volume mode @ zoom level `1.192` is mega, super slow. Especially the brush mode is laggy AF:
![volume](https://user-images.githubusercontent.com/1105056/38088908-e13d629a-335d-11e8-8e00-364190149946.gif)
- It doesn't work with Safari 11.0.3 for Mac. In fact it completely crashes the process:
![image](https://user-images.githubusercontent.com/1105056/38089347-70b06c96-335f-11e8-9a2e-e30b96ebe518.png)

hotzenklotz commented:
@philippotto You mentioned that there are changes to the bucket loading system. Does this impact any of the following:
- loading /pre-fetching strategies
- required bandwidth

hotzenklotz commented:
I just opened the same dataset, at the same position , zoom etc with the old and new version. I did not move or do anything. Just opened the link and waited for the network tab to finish downloading:

Anisotropy Brach
- https://anisotropy.webknossos.xyz/annotations/Explorational/5abcea35360000fb00c62010#745,621,153,0,0.40
- 1118 Requests
- 208MB

Master Branch
- https://master.webknossos.xyz/annotations/Explorational/5abceb2d3600009d00e7cafe#745,621,153,0,0.40
-129 Requests
- 17MB

philippotto commented on app/assets/javascripts/oxalis/controller/scene_controller.js:
```diff
@@ -335,8 +352,13 @@ class SceneController {
     });
+    listenToStoreProperty(
+      storeState => storeState.datasetConfiguration.interpolation,
+      // Setting interpolation is expensive as THREE will re-render everything and also re-upload
```
Right, before my changes this was always done when the store changed, which is expensive. Now, I made the listener more concrete. However, the comment is obsolete anyway, since the interpolation is done in the shader now. I'll remove the comment.

philippotto commented on app/assets/javascripts/oxalis/geometries/materials/abstract_plane_material_factory.js:
```diff
@@ -24,32 +29,107 @@ export type ShaderMaterialOptionsType = {
   }
 
+  /* eslint-disable no-unused-vars */
+  attachTextures(textures: TextureMapType): void {}
```
Class Inheritance + Flow + Linting = The magic you see above ;) When we switch the arbitrary mode to GPU rendering, this will be gone.

philippotto commented on app/assets/javascripts/oxalis/geometries/materials/plane_material_factory.js:
```diff
@@ -132,49 +271,356 @@ vec3 hsv_to_rgb(vec4 HSV)
+vec4 getRgbaAtXYIndex(sampler2D textures[textureCountPerLayer], float textureIdx, float textureWidth, float x, float y) {
+  vec2 accessPoint = (floor(vec2(x, y)) + 0.5) / textureWidth;
+
+  // Since WebGL doesnt allow dynamic texture indexing, we use an exhaustive if-else-construct
```
Yes, with WebGL 2 we could directly index into an array of textures. However, I don't think that the if-branches in the shader are a bottleneck for performance.

daniel-wer commented:
I think (and @hotzenklotz's measurements support this) we need to be more careful with what data we're requesting and uploading to the GPU. As I understand, currently we always (request and) upload 16x16 buckets for each plane onto the GPU, regardless of whether only 3x3 buckets are visible, is that correct?
Unfortunately I can barely test the branch, because my laptop gets really laggy. This is probably because I only have 8GB of RAM, and my integrated Intel 5500 uses the RAM as its video memory as well :/

In the meantime I fixed the "Unhandled byte count" error.

philippotto commented:
> @philippotto You mentioned that there are changes to the bucket loading system. Does this impact any of the following:
>
> loading /pre-fetching strategies
> required bandwidth

Buckets of the next "worse" zoomstep are always pulled. I think this attributes to 28% more data being pulled.

> When zooming in I have white flash / flickering. Zooming out works, however.
flickering

> in some cases the data does not load "all the way" and shows just the cross section

I cannot reproduce these two issues. Which datasets are you loading? Any specific positions were this happens?

> I think (and @hotzenklotz's measurements support this) we need to be more careful with what data we're requesting and uploading to the GPU. As I understand, currently we always (request and) upload 16x16 buckets for each plane onto the GPU, regardless of whether only 3x3 buckets are visible, is that correct?

Yes, I can adapt the ping strategy and the texture bucket manager to only pull/move what's necessary. However, this won't change the worst-case scenario. Meaning, if you are at a zoomLevel which demands 480 voxels per dimension, 16 (or 17) buckets are going to be pulled and moved to the GPU.
So, @hotzenklotz's measured 1118 requests are indeed overkill for the provided zoom level, but it's not unrealistic that this is necessary for another zoom level (3 * 17**2 plus pre-fetching slices of active viewport + fallback buckets > 1000 buckets).

> Unfortunately I can barely test the branch, because my laptop gets really laggy. This is probably because I only have 8GB of RAM, and my integrated Intel 5500 uses the RAM as its video memory as well :/

This is not good news. How many textures and which texture size does your laptop support? Which dataset did you open? Is it already a problem with only one color layer (no segmentation)? Can you open the performance tools and find a culprit?
In TextureBucketManager.processWriterQueue you can tweak `maxBucketCommitsPerFrame` to another number. That way, computation is more (or less) spread out over multiple frames. However, if your perf problems are really due to RAM shortage this will not really change things, I guess.

For color layers, we could properly leverage RGBA channels. However, I'm not sure if this will really reduce RAM requirements by a factor of 4. If you feel like it, you could try that out. However, this will only work for color layers (e.g., not RGB layers or volume layers).

> In the meantime I fixed the "Unhandled byte count" error.

Awesome, thank you! I fixed the issue with the arbitrary mode and added code for validating dataset/tracing requirements against hardware specs.

philippotto commented:
@hotzenklotz If you have some time on your hands, feel free to re-test the issues you raised before. I think/hope (as I could not reproduce everything) that the most recent commits fixed most of them. Do you still feel like brushing is slow "af"? That's my main concern at the moment.

hotzenklotz commented:
@philippotto The new version already works a lot better. Especially tracing mode feels much smoother both in Chrome and FF. All datasets load fine and I had no trouble loading the data (no missing buckets etc). Volume mode performance is improved but not great yet. Movement and the "trace" (outline) tool work well. Brush mode is still slow. When I am zoomed in enough it is slow but ok, but when I zoom out further it becomes unusable.

I retested the dowloaded data / bandwidth requirements with the same dataset, position etc as above and every is nice again. 
![image](https://user-images.githubusercontent.com/1105056/38427577-e49ab322-39b9-11e8-92bf-1e98862024c6.png)

Here is what I noticed during my last test session:
- When I zoom out really really far wK breaks in JS land: 
![image](https://user-images.githubusercontent.com/1105056/38425636-b6205ac4-39b4-11e8-87e1-60016200d33f.png)
- I get a ThreeJS Warning (might be harmless):
![image](https://user-images.githubusercontent.com/1105056/38426308-a0515958-39b6-11e8-9e47-e0919c0e6a3f.png)
- In volume mode, after labelling for a while (big blue dot) when I switch between zoom steps my annotation is completely gone. In other words, I can only see my new cell in one zoom step but not the other. Curiously when I delete cells it "penetrates" through all zoom steps. See GIF:
![volume_mode](https://user-images.githubusercontent.com/1105056/38427021-6efab38e-39b8-11e8-826f-dac970ebc99f.gif)
- Safari works now, but even tracing mode is unuseably slow. Very laggy. As soon as I load Safari Dev Tools it crashes.

This is a big upgrade over the last version I tested. We are on the right track.
![track gif-downsized](https://user-images.githubusercontent.com/1105056/38427781-7663dcac-39ba-11e8-9a36-9cfc07742b5d.gif)

philippotto commented:
> In volume mode, after labelling for a while (big blue dot) when I switch between zoom steps my annotation is completely gone. In other words, I can only see my new cell in one zoom step but not the other. Curiously when I delete cells it "penetrates" through all zoom steps. See GIF:

Partly this might be due to the fact that we downsample buckets for rendering higher zoom steps now. Since two "z-slices" are "squished together" your labeling might be "lost" since the cell IDs from the next slice might win. However, the colors change (invert?) exactly at the border of your brushed area which seems fishy. I'll into it (as well as into the other stuff).

philippotto commented:
> When I zoom out really really far wK breaks in JS land:

I cannot reproduce this. Which dataset did you use and which mode?

> I get a ThreeJS Warning (might be harmless):

Interesting. I don't get this warning, but it should be harmless. At the moment, the vertex shader is shared between orthogonal and arbitrary mode. However, the vUv variable is only used in the fragment shader of the arbitrary mode. I'm not sure whether it's worth splitting up the vertex shader only to mitigate this warning.

daniel-wer commented:
Mapping support has been restored. Feel free to test it using our API, for example: `webknossos.apiReady(3).then( api => api.data.setMapping(api.data.getVolumeTracingLayerName(), [0,1,2]))` or if you have a dataset with a mapping available, using the global mappings object, for example:
```javascript
mappings.getAll() -> ["astrocyte", "axons", ...]
mappings.activate("axons")
```
As some of the mappings are >200MB, have some patience, the console will inform you about the progress (as well as print some timing information for debugging purposes for now).
Currently the tracing view is frozen while the mapping is being built and the textures are created, there may be a way to avoid that, but then enabling the mapping would take even longer, so I'm not sure whether that is desired.
Ideally we will show a spinner while the mapping is loading, but I suppose this can be done as part of another PR that makes mappings more accessible.

I also saw that there is a mapping info view that can be used to temporarily disable an active mapping and shows information about the mapped ids, that is only enabled in volume tracings.
We should think about enabling this view for every tracing that has a segmention layer in my opinion.

philippotto commented on app/assets/javascripts/libs/utils.js:
```diff
@@ -415,6 +424,58 @@ const Utils = {
+
+  median8(dataArray: Array<number>): number {
+    // Returns the median of an already *sorted* array of size 8 (e.g., with sortArray8)
+    return Math.round((dataArray[3] + dataArray[4]) / 2);
```
Right, but it's only used for color data. Segmentation uses the mode, which is why this should be alright.

philippotto commented on app/assets/javascripts/oxalis/geometries/materials/abstract_plane_material_factory.js:
```diff
@@ -5,15 +5,20 @@
 import Model from "oxalis/model";
 import type { DatasetLayerConfigurationType } from "oxalis/store";
 import { listenToStoreProperty } from "oxalis/model/helpers/listener_helpers";
+import shaderEditor from "oxalis/model/helpers/shader_editor";
```
As long as it doesn't cause any trouble, I'd leave it there. It proved to be quite helpful for development and debugging. Maybe we pimp it up even more if we feel it's worth it.

philippotto commented on app/assets/javascripts/oxalis/model.js:
```diff
@@ -492,5 +582,40 @@ export class OxalisModel {
 
+function adaptResolutions(dataset: APIDatasetType): [APIDatasetType, Array<Vector3>] {
+  const adaptedLayers = dataset.dataSource.dataLayers.map(dataLayer => {
+    // Todo: the back-end should deliver the resolutions numerically sorted
```
Good catch, #2463 should have resolved this. I'll delete this code.

philippotto commented on app/assets/javascripts/oxalis/model/binary/data_cube.js:
```diff
@@ -194,6 +203,9 @@ class DataCube {
     const cube = this.cubes[zoomStep];
     if (cube != null) {
       const { boundary } = cube;
+      if (x >= boundary[0] || y >= boundary[1] || z >= boundary[2] || x < 0 || y < 0 || z < 0) {
```
Oh yes, I didn't see the comment above. The check became necessary when I temporarily disabled boundary checks globally and the calculated indices weren't unique anymore. That was a hard-to-find bug which is why I added the sanity check here. However, I'll re-remove the check so that performance is not harmed :)

philippotto commented on app/assets/javascripts/oxalis/model/binary/texture_bucket_manager.js:
```diff
@@ -0,0 +1,283 @@
+    // back (via pop). This ensures that the newest bucket "wins" if there are
+    // multiple buckets for the same index.
+    this.writerQueue = _.uniqBy(this.writerQueue, el => el._index);
+    const maxBucketCommitsPerFrame = 30;
```
Yes, I determined it empirically. However, this obviously depends on the machine. I'll rewrite it so that the loop stays busy until 16ms have passed. We could play with the value of 16ms a bit, but I think that's a reasonable default.

philippotto replied:
I played a bit around with the value and committed the version with 16ms but I couldn't really see a big difference. Maybe you can also fiddle with it to see what works best on your hardware.

daniel-wer commented:
I did some more manual testing and fixed the issues I found right away, looks good to me now.
Do you have a todo list @philippotto?

philippotto commented on app/assets/javascripts/oxalis/geometries/materials/plane_material_factory.js:
```diff
@@ -651,7 +652,9 @@ void main() {
-    vec4 HSV = vec4( mod( id * golden_ratio, 1.0), 1.0, 1.0, 1.0 );
+    // To fight float imprecision, extract the smallest 8-bit first (mod with 256.0), then multiply with the golden golden_ratio
+    // and finally get a value between 0.0 and 1.0
+    vec4 HSV = vec4( mod( mod(id, 256.0) * golden_ratio, 1.0), 1.0, 1.0, 1.0 );
```
Nice find & fix!

philippotto commented:
Everything worked and felt really good :) The initial loading is a bit slow (it's unclear that something is done in the first place), which is why I added another console.log to indicate that it started. I'd suggest to just change the console.times to console.logs to have progress shown.

philippotto commented:
Since RGB data is rarely (at the moment there are not rgb datasets online?) used and the bug only affects usability when zooming out, I fixed the issue by simply not downsampling rgb buckets. A warning is shown for such datasets. We'll hear when/whether it's an important feature.

> Do you have a todo list @philippotto?

My todo list is empty now :) I'll give it to the lab for testing. iPad and older/weaker hardware should get special attention here.

philippotto commented:
I just found a new bug :(

- [ ] Switching between orthogonal and arbitrary modes, sometimes makes data disappear (skeletons are still rendered but the data textures are probably empty (?)).
